### PR TITLE
{Core} Add linux distribution info in telemetry

### LINKED ISSUE #24783: Add Linux distribution information in client telemetry
Currently the client telemetry logs the operating system (Windows, Linux, MacOS), platform and version via https://github.com/Azure/azure-cli/blob/060b414d8ac5e4d8f4fc17cc2e23c2154c762226/src/azure-cli-core/azure/cli/core/telemetry.py#L143-L145

Some example entries from client telemetry:

https://dataexplorer.azure.com/clusters/ddazureclients/databases/AzureCli?query=H4sIAAAAAAAAA52QMQ+CQAyFd39FNzAxF5ycMDHGnRDiYhwuUPSUu17uCqjxx3uKorNb8/r6ta+57DcdGvar27pRkzv0R3QIL61QGj1LbWEJ8kDxopqOhlz2a9JamgrSFKIOnVdkotDHC2NQi6tFSIHJs1PmEGeOLDpW6HdRSYaDTVRYy7Zh0XlRkkNBXnAYi/bTL2c7kP9AfW76pWWN5Jqc/gNn36MDb/jCEDLkb5RpL8/0LM8I8yRJQm0dnbDkT4TZuP0BN182/nUBAAA=
```
RawEventsAzCli
| where EventTimestamp > ago(7d)
| where RawCommand == 'version'
| extend Type = tostring(Properties['context.default.vs.core.os.type'])
| extend Version = tostring(Properties['context.default.vs.core.os.version'])
| extend Platform = tostring(Properties['context.default.vs.core.os.platform'])
| where Type == 'linux'
| take 1000
| project Version, Platform
```

```
| Version                                             | Platform                                                      |
| --------------------------------------------------- | ------------------------------------------------------------- |
| #58~20.04.1-ubuntu smp thu oct 13 13:09:46 utc 2022 | linux-5.15.0-52-generic-x86_64-with-glibc2.35                 |
| #1 smp thu feb 10 16:11:23 utc 2022                 | linux-4.18.0-365.el8.x86_64-x86_64-with-centos-8              |
| #1 smp fri apr 2 22:23:49 utc 2021                  | linux-5.10.16.3-microsoft-standard-wsl2-x86_64-with-glibc2.31 |
| #61-ubuntu smp fri oct 14 13:05:50 utc 2022         | linux-5.14.0-1054-oem-x86_64-with-glibc2.31                   |
| #1 smp sat oct 8 18:13:21 utc 2022                  | linux-3.10.0-1160.80.1.el7.x86_64-x86_64-with-glibc2.31       |
| #1237-microsoft sat sep 11 14:32:00 pst 2021        | linux-4.4.0-19041-microsoft-x86_64-with-glibc2.31             |
```

Even though it has words like `ubuntu` (from Version column), `el8`, `el7` (from Platform column), some entries don't tell which Linux distributions they are, such as the last entry.

If the underlying base engine is WSL, no Linux distribution information can be detected:

Ubuntu 22.04 on WSL2:

```
$ python3 -c "import platform; print(platform.platform()); print(platform.version())"
Linux-5.10.102.1-microsoft-standard-WSL2-x86_64-with-glibc2.35
#1 SMP Wed Mar 2 00:30:59 UTC 2022
```

Docker on WSL2:

```
> docker run -it --rm debian:11
# apt update; apt install --yes python3
# python3 -c "import platform; print(platform.platform()); print(platform.version())"
Linux-5.10.102.1-microsoft-standard-WSL2-x86_64-with-glibc2.31
#1 SMP Wed Mar 2 00:30:59 UTC 2022
```

```
> docker run -it --rm fedora
# python3 -c "import platform; print(platform.platform()); print(platform.version())"
Linux-5.10.102.1-microsoft-standard-WSL2-x86_64-with-glibc2.36
#1 SMP Wed Mar 2 00:30:59 UTC 2022
```

# Possible solution

Instead of relying on Python's `platform` module, an industry standard is to provide distribution information in `/etc/os-release`.

Debian

```
> docker run -it --rm debian:11 cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
NAME="Debian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```

Ubuntu

```
> docker run -it --rm ubuntu:22.04 cat /etc/os-release
PRETTY_NAME="Ubuntu 22.04.1 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.1 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy
```

Fedora

```
> docker run -it --rm fedora cat /etc/os-release
NAME="Fedora Linux"
VERSION="37 (Container Image)"
ID=fedora
VERSION_ID=37
VERSION_CODENAME=""
PLATFORM_ID="platform:f37"
PRETTY_NAME="Fedora Linux 37 (Container Image)"
ANSI_COLOR="0;38;2;60;110;180"
LOGO=fedora-logo-icon
CPE_NAME="cpe:/o:fedoraproject:fedora:37"
DEFAULT_HOSTNAME="fedora"
HOME_URL="https://fedoraproject.org/"
DOCUMENTATION_URL="https://docs.fedoraproject.org/en-US/fedora/f37/system-administrators-guide/"
SUPPORT_URL="https://ask.fedoraproject.org/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"
REDHAT_BUGZILLA_PRODUCT="Fedora"
REDHAT_BUGZILLA_PRODUCT_VERSION=37
REDHAT_SUPPORT_PRODUCT="Fedora"
REDHAT_SUPPORT_PRODUCT_VERSION=37
VARIANT="Container Image"
VARIANT_ID=container
```

Red Hat

```
> docker run -it --rm registry.access.redhat.com/ubi8/ubi cat /etc/os-release
NAME="Red Hat Enterprise Linux"
VERSION="8.7 (Ootpa)"
ID="rhel"
ID_LIKE="fedora"
VERSION_ID="8.7"
PLATFORM_ID="platform:el8"
PRETTY_NAME="Red Hat Enterprise Linux 8.7 (Ootpa)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:redhat:enterprise_linux:8::baseos"
HOME_URL="https://www.redhat.com/"
DOCUMENTATION_URL="https://access.redhat.com/documentation/red_hat_enterprise_linux/8/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"

REDHAT_BUGZILLA_PRODUCT="Red Hat Enterprise Linux 8"
REDHAT_BUGZILLA_PRODUCT_VERSION=8.7
REDHAT_SUPPORT_PRODUCT="Red Hat Enterprise Linux"
REDHAT_SUPPORT_PRODUCT_VERSION="8.7"
```

openSUSE

```
> docker run -it --rm opensuse/leap cat /etc/os-release
NAME="openSUSE Leap"
VERSION="15.4"
ID="opensuse-leap"
ID_LIKE="suse opensuse"
VERSION_ID="15.4"
PRETTY_NAME="openSUSE Leap 15.4"
ANSI_COLOR="0;32"
CPE_NAME="cpe:/o:opensuse:leap:15.4"
BUG_REPORT_URL="https://bugs.opensuse.org"
HOME_URL="https://www.opensuse.org/"
DOCUMENTATION_URL="https://en.opensuse.org/Portal:Leap"
LOGO="distributor-logo-Leap"
```

### PR DESCRIPTION
**Pull request by evelyn-ys**
**Description**

Fix #24783 

Currently we use `platform.system()`, `platform.version()` and `platform.platform()` to log OS info in telemetry https://github.com/Azure/azure-cli/blob/060b414d8ac5e4d8f4fc17cc2e23c2154c762226/src/azure-cli-core/azure/cli/core/telemetry.py#L143-L145

But it's not clear in some scenarios for linux distribution like @jiasli mentioned in the linked issue.
In this PR, we import `distro` lib for linux and log detailed info. Another two fields will be added to `Properties` in telemetry:
- `Context.Default.VS.Core.Distro.Name`
- `Context.Default.VS.Core.Distro.Version`

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

File: src/azure-cli-core/azure/cli/core/telemetry.py [hunk 1]
```diff
@@ -556,6 +559,24 @@ def _get_hash_machine_id():
...
+
+
+@decorators.suppress_all_exceptions(fallback_return='')
+def _get_distro_version():
+    try:
+        import distro
+        return distro.version(pretty=True)
+    except ImportError:
+        return ''
+
```
Comment (bebound) [inline src/azure-cli-core/azure/cli/core/telemetry.py:578]: Should we add `pretty=True`？
If it is added, name and version is `Ubuntu 22.04.1 LTS`, `22.04 (jammy)`.
Without it, they become `Ubuntu`, `22.04`.
I think the later one is easy to use if you want to get aggregate data.

Comment (evelyn-ys) [reply src/azure-cli-core/azure/cli/core/telemetry.py:578]: I added `pretty=True` to collect as much data as possible. But I'm also open for the short option according to how we really need to use the info.

@jiasli What do you think?

Comment (evelyn-ys) [reply src/azure-cli-core/azure/cli/core/telemetry.py:578]: Maybe we can keep `pretty=True` for `distro.name` but omit for `distro.version`?

**Comment by yonzhan** [comment]
Add linux distribution info

File: src/azure-cli-core/azure/cli/core/telemetry.py (hunk 1 above)
Comment (jiasli) [reply src/azure-cli-core/azure/cli/core/telemetry.py:578]:
```
$ python
>>> import distro
>>> distro.name()
'Ubuntu'
>>> distro.name(pretty=True)
'Ubuntu 20.04.4 LTS'
>>> distro.version()
'20.04'
>>> distro.version(pretty=True)
'20.04 (focal)'
>>> distro.id()
'ubuntu'
```

The above information is retrieved from

```
$ cat /etc/os-release
NAME="Ubuntu"
VERSION="20.04.4 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.4 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
```

Comment (jiasli) [reply src/azure-cli-core/azure/cli/core/telemetry.py:578]: According to the source code os `distro`:

https://github.com/python-distro/distro/blob/6e41b8d5b0f3b962c7fcfff37271fcb4398c2558/src/distro/distro.py#L887

```py
        versions = [
            self.os_release_attr("version_id"),
```

`version_id` is the most preferred result for `distro.version()`.

File: src/azure-cli-core/azure/cli/core/telemetry.py [hunk 2]
```diff
@@ -556,6 +559,24 @@ def _get_hash_machine_id():
     return str(uuid.UUID(_get_hash_mac_address()[0:32]))
 
 
+@decorators.suppress_all_exceptions(fallback_return='')
+def _get_distro_name():
+    try:
+        import distro
+        return distro.name(pretty=True)
+    except ImportError:
+        return ''
```
Comment (jiasli) [inline src/azure-cli-core/azure/cli/core/telemetry.py:568]: Do we really need this try except clause given all expressions are already suppressed by `decorators.suppress_all_exceptions`?

Comment (evelyn-ys) [reply src/azure-cli-core/azure/cli/core/telemetry.py:568]: Without this try except, IDE will remind error
image

But with try except, it only reminds warning
image

I don't like the error reminder so...😂

**Review by jiasli — APPROVED** [review]

**Pull request merged**